[rum] guide on upgrade to browser sdk v3 (#11257)
DataDog/documentation: content/en/real_user_monitoring/guide/v3-browser-sdk-upgrade.md

> ---

Reviewer: Rename file to `browser-sdk-upgrade.md`

> title: Upgrading the Browser SDK to v3

Reviewer: ```suggestion
title: Upgrade the Browser SDK
```

kind: guide
further_reading:
- link: '/real_user_monitoring/explorer'
  tag: 'Documentation'
  text: 'Visualize your RUM data in the Explorer'
- link: "https://www.datadoghq.com/blog/session-replay-datadog/"
  tag: "Blog"
  text: "Use Datadog Session Replay to view real-time user journeys"
---

## Overview

> Version 3 of the RUM Browser SDK introduces [Session Replay][1]. With this new major version, several breaking changes were made to the Browser SDK.

Reviewer: ```suggestion
RUM Browser SDK v3 introduces [Session Replay][1]. With this major version update, several breaking changes were made to the Browser SDK.
```

> ### Stopping the generation of a RUM error for failed network requests

Reviewer: ```suggestion
## Changes
### RUM errors
```

The Browser SDK no longer issues [RUM errors][2] for failed XHR and Fetch calls. These failed network requests are still collected as [RUM resources][3], which contain the status code attribute.

> If you want to continue seeing these failed network requests as RUM errors, Datadog recommends intercepting the resource with the [beforeSend API][4], checking the `status_code` property and manually sending an error with the [addError API][5].

Reviewer: ```suggestion
To continue seeing the failed network requests as RUM errors, Datadog recommends intercepting the resource with the [beforeSend API][4], checking the `status_code` property, and manually sending an error with the [addError API][5].
```

```javascript
beforeSend: (event) => {
    if (event.type === 'resource' && event.resource.status_code >= 500) {
        datadogRum.addError(`${event.resource.method} ${event.resource.url} ${event.resource.status_code}`); // "GET https://www.example.com/ 504"
    }
}
```

> ### Removing the source attribute from the addError API

Reviewer: ```suggestion
### Source attribute
```

> The Browser SDK no longer lets you specify the source of an error collected with the [addError API][5]. All errors collected with this API now have their source attribute set to `custom`. The [addError API][5] accepts a context object as its second parameter, which should be used to pass extra context about the error.

Reviewer: ```suggestion
The Browser SDK no longer lets you specify the source of an error collected with the [addError API][5]. All errors collected with this API have their source attribute set to `custom`. The [addError API][5] accepts a context object as its second parameter, which should be used to pass extra context about the error.
```

> ### Deprecated Typescript types

Reviewer: ```suggestion
## Deprecated
### TypeScript types
```

| Old types                    | New types                    |
| ---------------------------- | ---------------------------- |
| RumUserConfiguration         | RumInitConfiguration         |
| RumRecorderUserConfiguration | RumRecorderInitConfiguration |

> ### Deprecated initialization options

Reviewer: ```suggestion
### Initialization options
```

| Old options        | New options |
| ------------------ | ----------- |
| publicApiKey       | clientToken |
| datacenter         | site        |
| resourceSampleRate | NONE        |

> ### Deprecated API

Reviewer: ```suggestion
### API
```
